src/Microsoft.AspNetCore.OData/Batch/ODataBatchReaderExtensions.cs

```diff
@@ -140,9 +146,19 @@ public static class ODataBatchReaderExtensions
 
             foreach (var header in batchRequest.Headers)
             {
-                // Copy headers from batch, overwriting any existing headers.
                 string headerName = header.Key;
                 string headerValue = header.Value;
+
+                if (headerName.Trim().ToLowerInvariant() == "prefer")
+                {
+                    // in the case of Prefer header, we don't want to overwrite,
+                    // instead we merge preferences defined in the individual request with those inherited from the batch
+                    request.Headers.TryGetValue(headerName, out StringValues batchReferences);
+                    request.Headers[headerName] = MergeIndividualAndBatchPreferences(headerValue, batchReferences);
+                    continue;
+                }
```

Comment:
@habbes Out of curiosity, are the headers and preferences finite and known and do we know for sure that only headers: `content-length` and `content-type` and preferences: `respond-async`, `continue-on-error` and `odata.continue-on-error` are not inheritable from parent (or not applicable to nested requests?

Reply:
These are what we decided not to inherit following a discussion in the issue that this PR addressed. Whether the possible headers and preferences are finite, I'm not sure what you mean by that. It's possible that users can define custom headers or preferences, or that new headers or preferences may be introduced into the spec over time. I don't know whether this is what you mean by finite.

Reply:
@habbes If user can define headers and preferences, the question would be whether the nested requests inheriting those from the batch is okay. It'd be hard to know though. The outcome of inheriting or not inheriting the custom headers and preferences can swing one way or the other

Reply:
@gathogojr yes it would be hard. But if they are custom headers, then maybe it should be up to the user to decide to place them on the individual requests that they apply to if they don't want them to be copied to all other requests in the batch.

Reply:
Good question. Unfortunately, there is no easy way to determine whether or not a preference is inherited by individual requests within a batch -- it depends on the definition of the header.  For the preference headers that OData defines (or defines semantics for) we explicitly call out what should be inherited and what should not be, which is what Clement has encoded here.  Unfortunately, adding additional preferences in the future will require that we update this code appropriately.  The safe default for unknown preferences is to copy them over and let the individual statement have the option of handling.

---
In reply to: [386260180](https://github.com/OData/WebApi/pull/2064#discussion_r386260180) [](ancestors = 386260180)

Reply:
Exactly. Unfortunately, we don't know the intended behavior of custom headers and preferences, so the default is to copy them over and let the individual statement handle.

---
In reply to: [386295560](https://github.com/OData/WebApi/pull/2064#discussion_r386295560) [](ancestors = 386295560)